ament_cmake_vendor_package/cmake/ament_vendor.cmake

```diff
@@ -0,0 +1,265 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+macro(ament_vendor TARGET_NAME)
+  if(NOT PROJECT_NAME)
+    message(FATAL_ERROR "ament_vendor() must be called after project()")
+  endif()
+
+  if(_${PROJECT_NAME}_AMENT_PACKAGE)
+    message(FATAL_ERROR "ament_vendor() must be called before ament_package()")
+  endif()
+
+  cmake_parse_arguments(_ARG "GLOBAL_HOOK;SKIP_INSTALL" "SOURCE_SUBDIR;VCS_TYPE;VCS_URL;VCS_VERSION;SATISFIED" "CMAKE_ARGS;PATCHES" ${ARGN})
+  if(_ARG_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "ament_vendor() called with unused arguments: "
+      "${_ARG_UNPARSED_ARGUMENTS}")
+  endif()
+
+  if(NOT _ARG_VCS_URL)
+    message(FATAL_ERROR "ament_vendor() must be called with the VCS_URL argument")
+  endif()
+  if(_ARG_VCS_TYPE STREQUAL "path")
+    if(_ARG_SOURCE_SUBDIR)
+      message(FATAL_ERROR "ament_vendor() cannot use VCS_TYPE 'path' with SOURCE_SUBDIR argument")
+    endif()
+    if(_ARG_PATCHES)
+      message(FATAL_ERROR "ament_vendor() cannot use VCS_TYPE 'path' with PATCHES argument")
```

Comment:
The `path` type is special because it bypasses the download step entirely. Any other value is passed to `vcstool`, so all of these are supported: https://github.com/dirk-thomas/vcstool/tree/master/vcstool/clients

The reason that `PATCHES` isn't allowed with the `path` type is that the sources are used directly from the source tree, so patching them would make changes to the original repository. It doesn't make sense to include the patches there anyway - just change the files directly.